### PR TITLE
Decimal fast aggregations

### PR DESCRIPTION
**Pull request by pnowojski**
Benchmark results with old code (@wbiela pull request with avg/sum aggregations):

```
Benchmark                                                                                                             (project)          (type)  Mode  Cnt     Score    Error  Units
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, avg(totalprice)   decimal(14,2)  avgt   30   274.957 ±  6.592  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, avg(totalprice)  decimal(30,10)  avgt   30   439.031 ±  7.239  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, min(totalprice)   decimal(14,2)  avgt   30   206.957 ±  5.151  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, min(totalprice)  decimal(30,10)  avgt   30   535.220 ± 28.227  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash  orderstatus, sum(totalprice), avg(totalprice), min(totalprice), max(totalprice)   decimal(14,2)  avgt   30   412.677 ±  9.621  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash  orderstatus, sum(totalprice), avg(totalprice), min(totalprice), max(totalprice)  decimal(30,10)  avgt   30  1165.726 ± 82.984  ms/op
```

After changes:

```
Benchmark                                                                                                             (project)  (type)  Mode  Cnt    Score    Error  Units
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, avg(totalprice)          double  avgt   10  166.982 ±  1.467  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, min(totalprice)          double  avgt   10  149.677 ±  9.512  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash  orderstatus, sum(totalprice), avg(totalprice), min(totalprice), max(totalprice)          double  avgt   10  174.185 ± 5.591  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, avg(totalprice)   decimal(14,2)  avgt   10  172.849 ± 12.891  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, avg(totalprice)  decimal(30,10)  avgt   10  231.429 ±  3.415  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, min(totalprice)   decimal(14,2)  avgt   10  151.328 ±  7.758  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash                                                     orderstatus, min(totalprice)  decimal(30,10)  avgt   10  222.131 ±  7.071  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash  orderstatus, sum(totalprice), avg(totalprice), min(totalprice), max(totalprice)   decimal(14,2)  avgt   10  251.888 ± 19.230  ms/op
BenchmarkDecimalAggregation.benchmarkBuildHash  orderstatus, sum(totalprice), avg(totalprice), min(totalprice), max(totalprice)  decimal(30,10)  avgt   10  341.743 ±  4.464  ms/op
```

### CODE REVIEW

**Comment by pnowojski** [comment]
Please review only 3 commits as a part of this pull request:
- Fix add overflow in unscaled decimal
- Use fast decimal in SUM aggregation
- Use fast decimal in AVG aggregation

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 1]
```diff
@@ -301,11 +301,11 @@ private static void addUnsigned(Slice left, Slice right, Slice result, boolean r
 
         product = (l3 & LONG_MASK) + (r3 & LONG_MASK) + (product >> 32);
 
-        int z3 = (int) product;
+        int z3 = (int) product & (~SIGN_INT_MASK);
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:325]: move below overflow check and remove this &

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            return unscaledDecimals.get(getGroupId());
+        }
+
+        @Override
+        public void setLongDecimal(Slice value)
+        {
+            requireNonNull(value, "value is null");
+            if (getLongDecimal() == null) {
+                numberOfElements += 1;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java]: use `++`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowState.java [hunk 3]
```diff
@@ -15,14 +15,17 @@
...
-@AccumulatorStateMetadata(stateFactoryClass = BigIntegerStateFactory.class, stateSerializerClass = BigIntegerStateSerializer.class)
-public interface BigIntegerState
+@AccumulatorStateMetadata(stateFactoryClass = LongDecimalWithOverflowStateFactory.class, stateSerializerClass = LongDecimalWithOverflowStateSerializer.class)
+public interface LongDecimalWithOverflowState
         extends AccumulatorState
 {
-    BigInteger getBigInteger();
+    Slice getLongDecimal();
+
+    void setLongDecimal(Slice unscaledDecimal);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowState.java]: shouldn't you set both long decimal and overflow at the same time? It's strange that you can set them separately.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java [hunk 4]
```diff
@@ -33,21 +34,33 @@ public Type getSerializedType()
...
-    public void serialize(BigIntegerState state, BlockBuilder out)
+    public void serialize(LongDecimalWithOverflowState state, BlockBuilder out)
     {
-        if (state.getBigInteger() == null) {
+        if (state.getLongDecimal() == null) {
             out.appendNull();
         }
         else {
-            VARBINARY.writeSlice(out, Slices.wrappedBuffer(state.getBigInteger().toByteArray()));
+            byte[] longDecimalBytes = state.getLongDecimal().getBytes();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java]: this copies bytes. I think you can directly write long decimal to out.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java [hunk 5]
```diff
@@ -33,21 +34,33 @@ public Type getSerializedType()
...
+
+            Slice slice = Slices.allocate(Long.BYTES + longDecimalBytes.length);
+            SliceOutput output = slice.getOutput();
+
+            output.writeLong(state.getOverflow());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java]: you could just use `VARBINARY.writeSlice(out, Slices.wrappedLongArray(state.getOverflow())`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java [hunk 6]
```diff
@@ -33,21 +34,33 @@ public Type getSerializedType()
...
     @Override
-    public void deserialize(Block block, int index, BigIntegerState state)
+    public void deserialize(Block block, int index, LongDecimalWithOverflowState state)
     {
         if (!block.isNull(index)) {
-            state.setBigInteger(new BigInteger(VARBINARY.getSlice(block, index).getBytes()));
+            SliceInput slice = VARBINARY.getSlice(block, index).getInput();
+
+            state.setOverflow(slice.readLong());
+            byte[] longDecimalBytes = slice.readSlice(slice.available()).getBytes();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java]: why do you need `byte[]` intermediate array. I think you can directly put read slice in state

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+{
+    private static final BigInteger TWO = new BigInteger("2");
+    private static final DecimalType TYPE = createDecimalType(38, 0);
+
+    private LongDecimalWithOverflowState state;
+
+    @BeforeMethod
+    public void setUp()
+    {
+        state = new LongDecimalWithOverflowStateFactory().createSingleState();
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java:42]: what about `GroupedLongDecimalWithOverflowState`?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 8]
```diff
@@ -226,9 +226,27 @@ public static Slice add(Slice left, Slice right)
...
+        }
+    }
+
+    /**
+     * Instead of throwing overflow exception, this function returns:
+     *     0 when there was no overflow
+     *     +1 when there was overflow
+     *     -1 when there was underflow
+     */
+    public static long addNoThrow(Slice left, Slice right, Slice result)
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: Maybe `addUnsafe` or `addReturnOverflow`?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 9]
```diff
@@ -272,9 +293,9 @@ else if (compare < 0) {
     }
 
     /**
-     * This method ignores signs of the left and right.
+     * This method ignores signs of the left and right. Returns overflow value.
      */
-    private static void addUnsigned(Slice left, Slice right, Slice result, boolean resultNegative)
+    private static long addUnsigned(Slice left, Slice right, Slice result, boolean resultNegative)
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: `addUnsignedUnsafe` or `addUnsignedReturnOverflow`?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java [hunk 10]
```diff
@@ -181,10 +186,25 @@ public static void outputLongDecimal(DecimalType type, BigIntegerAndLongState st
...
+
+        if (state.getOverflow() != 0) {
+            BigInteger overflowMultiplier = new BigInteger("2").pow(UNSCALED_DECIMAL_128_SLICE_LENGTH * 8 - 1);
+            BigInteger overflow = overflowMultiplier.multiply(BigInteger.valueOf(state.getOverflow()));
+            sum = sum.add(new BigDecimal(overflow));
+        }
         return sum.divide(count, type.getScale(), ROUND_HALF_UP);
     }
+
+    private static void accumulateAndCheckOverflow(Slice unscaledValue, LongDecimalWithOverflowAndLongState state)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: `accumulateAndUpdateOverflow`?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java [hunk 11]
```diff
@@ -181,10 +186,25 @@ public static void outputLongDecimal(DecimalType type, BigIntegerAndLongState st
...
-    private static BigDecimal average(BigIntegerAndLongState state, DecimalType type)
+    @VisibleForTesting
+    public static BigDecimal average(LongDecimalWithOverflowAndLongState state, DecimalType type)
     {
-        BigDecimal sum = new BigDecimal(state.getBigInteger(), type.getScale());
+        BigDecimal sum = new BigDecimal(Decimals.decodeUnscaledValue(state.getLongDecimal()), type.getScale());
         BigDecimal count = BigDecimal.valueOf(state.getLong());
+
+        if (state.getOverflow() != 0) {
+            BigInteger overflowMultiplier = new BigInteger("2").pow(UNSCALED_DECIMAL_128_SLICE_LENGTH * 8 - 1);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: this should be static variable

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: you can use `BigInteger.shiftLeft`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateFactory.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+
+    public static class SingleLongDecimalWithOverflowAndLongState
+            extends LongDecimalWithOverflowStateFactory.SingleLongDecimalWithOverflowState
+            implements LongDecimalWithOverflowAndLongState
+    {
+        public static final int SIZE = SIZE_OF_LONG * 2 + UNSCALED_DECIMAL_128_SLICE_LENGTH;
+
+        protected long longValue;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateFactory.java:87]: what is this long, a counter? Then maybe name it `counter`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 13]
```diff
@@ -36,29 +34,35 @@ public Type getSerializedType()
...
-            Slice slice = Slices.allocate(Long.BYTES + bigIntegerBytes.length);
+            byte[] longDecimalBytes = state.getLongDecimal().getBytes();
+
+            Slice slice = Slices.allocate(Long.BYTES + Long.BYTES + longDecimalBytes.length);
             SliceOutput output = slice.getOutput();
+
             output.writeLong(state.getLong());
-            output.writeBytes(bigIntegerBytes);
+            output.writeLong(state.getOverflow());
+            output.writeBytes(longDecimalBytes);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: ditto, I think you can directly write long decimal to out

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 14]
```diff
@@ -36,29 +34,35 @@ public Type getSerializedType()
...
+    public void deserialize(Block block, int index, LongDecimalWithOverflowAndLongState state)
     {
         if (!block.isNull(index)) {
             SliceInput slice = VARBINARY.getSlice(block, index).getInput();
+
             state.setLong(slice.readLong());
-            byte[] bigIntegerBytes = slice.readSlice(slice.available()).getBytes();
-            state.setBigInteger(new BigInteger(bigIntegerBytes));
+            state.setOverflow(slice.readLong());
+            byte[] longDecimalBytes = slice.readSlice(slice.available()).getBytes();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: ditto, maybe long decimal bytes is not needed

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Test
+    public void testOverflow()
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java:46]: add test that there is an exception when you try to return final value with overflow

**Comment by sopel39** [comment]
lgtm, comments

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java (hunk 2 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 1 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:325]: this is required in next commit

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowState.java (hunk 3 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowState.java]: it seems like a convention for states to have only simple setters/getters. And actually it is more convenient that way

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java (hunk 4 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java]: If we don't copy, we will be blocking GC from freeing serialized channel memory (which contains also overflow and count)

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java (hunk 5 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java]: I have to write one single `VARBINARY` to the `BlockBuilder`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java (hunk 6 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java]: read slices would consist only of 50% of the incoming channel so this would give us huge memory leak.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java (hunk 7 above)
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java:42]: Covered by `TestLongAverage/SumAggregation` classes. This is just unit test for arithmetic.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 8 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 9 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java (hunk 10 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java (hunk 11 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateFactory.java (hunk 12 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateFactory.java:87]: It's carried from @ilfrin PR. I remember raiding same issue there and it was rejected :(

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java (hunk 15 above)
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestDecimalSumAggregation.java:46]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java (hunk 13 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java (hunk 14 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: as earlier copy is required
